internal/cephfs/core/quiesce.go

```diff
@@ -0,0 +1,237 @@
+	mapping map[string][]string,
+	conn *util.ClusterConnection,
+) (FSQuiesceClient, error) {
+	fsa, err := conn.GetFSAdmin()
+	if err != nil {
+		return nil, err
+	}
+
+	return &fsQuiesce{
+		fsName:                fsName,
+		volumes:               volumes,
+		subVolumeGroupMapping: mapping,
+		fsa:                   fsa,
+	}, nil
+}
+
+// GetVolumes returns the list of volumes in the filesystem that are to be
+// quiesced.
+func (fq *fsQuiesce) GetVolumes() []Volume {
+	return fq.volumes
+}
+
+// getMembers returns the list of names in the format
+// group/subvolume that are to be quiesced. This is the format that the
+// ceph fs quiesce expects.
+// Example: ["group1/subvolume1", "group1/subvolume2", "group2/subvolume1"].
+func (fq *fsQuiesce) getMembers() []string {
+	volName := []string{}
+	for svg, sb := range fq.subVolumeGroupMapping {
+		for _, s := range sb {
+			name := svg + "/" + s
+			volName = append(volName, name)
+		}
+	}
+
+	return volName
+}
+
+func (fq *fsQuiesce) FSQuiesce(
+	ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
+	}
+	log.DebugLog(ctx,
+		"FSQuiesce for reserveName %s: members:%v options:%v",
+		reserveName,
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
+		qInfo := resp.Sets[reserveName]
+
+		return &qInfo, nil
+	}
+
+	log.ErrorLog(ctx, "failed to quiesce filesystem %s", err)
+
+	return nil, err
+}
+
+func (fq *fsQuiesce) FSQuiesceWithExpireTimeout(ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
+	}
+	log.DebugLog(ctx,
+		"FSQuiesceWithExpireTimeout for reserveName %s: members:%v options:%v",
+		reserveName,
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
+		qInfo := resp.Sets[reserveName]
+
+		return &qInfo, nil
+	}
+
+	log.ErrorLog(ctx, "failed to quiesce filesystem with expire timeout %s", err)
+
+	return nil, err
+}
+
+func (fq *fsQuiesce) ResetFSQuiesce(ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Reset:      true,
+		AwaitFor:   0,
+		Timeout:    180,
+		Expiration: 180,
+	}
+	// Reset the filesystem quiesce so that the timer will be reset, and we can
+	// reuse the same reservation if it has already failed or timed out.
+	log.DebugLog(ctx,
+		"ResetFSQuiesce for reserveName %s: members:%v options:%v",
+		reserveName,
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
+		qInfo := resp.Sets[reserveName]
+
+		return &qInfo, nil
+	}
+
+	log.ErrorLog(ctx, "failed to reset timeout for quiesce filesystem %s", err)
+
+	return nil, err
```

Comment:
```suggestion
	if err != nil{
		log.ErrorLog(ctx, "failed to reset timeout for quiesce filesystem %s", err)

		return nil, err		
	}
	if resp == nil {
		return nil, fmt.Errorf("response is nil")
	}
	qInfo := resp.Sets[reserveName]

	return &qInfo, nil
```

This may lead to a panic in case both error and resp is nil ? 
I think its better to check both just in case ?

Reply:
IMHo I don't really see a case when both can be nil, I expect at least one to be set, if it's really required I can make that change